snuba/utils/concurrent.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+from concurrent.futures import Future
+from contextlib import contextmanager
+from threading import Lock, Thread
+from typing import Callable, Generic, Iterator, Optional, TypeVar
+
+
+T = TypeVar("T")
+
+
+def execute(
+    function: Callable[[], T], name: Optional[str] = None, daemon: Optional[bool] = None
+) -> Future[T]:
+    """
+    Execute a callable in a background thread, returning a ``Future`` object
+    that can be used to wait for the callable to exit (or check if it is
+    still running.)
+
+    This is similar to using ``concurrent.futures.ThreadPoolExecutor.submit``
+    but does not require a thread pool. Since each invocation will create
+    (and eventually shut down) new thread, this function is better suited for
+    long-lived background tasks (and generally IO bound ones, at that) than
+    distributing large workloads.
+    """
+    future: Future[T] = Future()
+
+    def run() -> None:
+        if not future.set_running_or_notify_cancel():
+            return
```

Comment:
When would this return False ?
Do you need to explicitly set the future to running state since you do not rely on a thread pool ?

Reply:
> When would this return False ?

It's extremely unlikely but not impossible that [`Future.cancel`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.cancel) is called between the time that the future is instantiated, and the thread starts.

> Do you need to explicitly set the future to running state since you do not rely on a thread pool ?

Yeah, kind of. This is normally handled by the executor implementation: https://github.com/python/cpython/blob/8289e27393395ee903bd096d42e07c112d7f15c6/Lib/concurrent/futures/thread.py#L57-L58